ENIGMAsystem/SHELL/Collision_Systems/Polygon/Polygonfuncs.cpp

```diff
@@ -0,0 +1,764 @@
+            enigma::get_bbox_border(left, top, right, bottom, inst);
+
+            // Sweep and Prune Check
+            if ((left <= (rleft + rwidth) && rleft <= right && top <= (rtop + rheight) && rtop <= bottom) == inside) 
+            {
+                inst->deactivate();
+                enigma::instance_deactivated_list.insert(inode_pair(inst->id,inst));
+            }
+        }
+    }
+
+    void instance_activate_region(int rleft, int rtop, int rwidth, int rheight, bool inside) 
+    {
+        // Iterating over the instances
+        std::map<int,enigma::object_basic*>::iterator iter = enigma::instance_deactivated_list.begin();
+        while (iter != enigma::instance_deactivated_list.end()) 
+        {
+            enigma::object_collisions* const inst = (enigma::object_collisions*) iter->second;
+
+            // no sprite/mask/polygon then no collision
+            if (inst->sprite_index == -1 && inst->mask_index == -1 && inst->polygon_index == -1)
+            {
+                ++iter;
+                continue;
+            }
+
+            // Bounding Box retrieval and collision check
+            int left, top, right, bottom;
+            enigma::get_bbox_border(left, top, right, bottom, inst);
+
+            if ((left <= (rleft + rwidth) && rleft <= right && top <= (rtop + rheight) && rtop <= bottom) == inside) 
+            {
+                inst->activate();
+                enigma::instance_deactivated_list.erase(iter++);
+            } 
+            else 
+            {
+                ++iter;
+            }
+        }
+    }
+
+}
+
+static bool line_ellipse_intersects(cs_scalar rx, cs_scalar ry, cs_scalar x, cs_scalar ly1, cs_scalar ly2)
+{
+    // Formula: x^2/a^2 + y^2/b^2 = 1   <=>   y = +/- sqrt(b^2*(1 - x^2/a^2))
+
+    const cs_scalar inner = ry*ry*(1 - x*x/(rx*rx));
+    if (inner < 0) {
+        return false;
+    }
+    else {
+        const cs_scalar y1 = -sqrt(inner), y2 = sqrt(inner);
+        return y1 <= ly2 && ly1 <= y2;
+    }
+}
+
+namespace enigma_user
+{
+
+    void instance_deactivate_circle(int x, int y, int r, bool inside, bool notme)
+    {
+        // Iterating over the instances
+        for (enigma::iterator it = enigma::instance_list_first(); it; ++it)
+        {
+            // boolean given notme than collision not checked for this instance
+            if (notme && (*it)->id == enigma::instance_event_iterator->inst->id)
+                continue;
+
+            // Fetching the instance
+            enigma::object_collisions* const inst = (enigma::object_collisions*) *it;
+
+            // If no sprite/mask/polygon then no collision
+            if (inst->sprite_index == -1 && (inst->mask_index == -1)) 
+                continue;
+
+            // Doing a Sweep and Prune check using BBOX
+            int left, top, right, bottom;
+            enigma::get_bbox_border(left, top, right, bottom, inst);
+
+            // TODO (Nabeel) : Does this needs a polygon collision check?
```

Comment:
I'm not sure about this one, myself, but I would assume so. These functions are terribly monotonous; this might be a place to use a template to handle collision checking an arbitrary instance, so that instance iteration logic (for activated and deactivated instances alike) can be reused. For example, you could pass a class that contains x, y, and radius and calls `collision_circle` in `operator()` to handle `instance_deactivate_circle`. I'm happy to help you with the template magic as needed.

Reply:
Converting these functions to more generalizable will improve code consistency and will decrease the redundancy that we are seeing here. This links back to the collision generalization issue that I mentioned earlier. I would love to have some template usage here